src/main/java/com/cryptomorin/xseries/particles/ParticleDisplay.java

```diff
@@ -268,8 +303,25 @@ public static ParticleDisplay edit(@Nonnull ParticleDisplay display, @Nonnull Co
                         // Size
                         (colors.length > 3 ? NumberUtils.toFloat(colors[3]) : 1.0f)};
             }
+        } else if (blockdata != null) {
+            Material material = Material.getMaterial(blockdata);
+            if (material != null && material.isBlock()) {
+                display.data = material.createBlockData();
+            }
+        } else if (item != null) {
+            Material material = Material.getMaterial(item);
+            if (material != null && material.isItem()) {
+                display.data = new ItemStack(material, 1);
+            }
+        } else if (materialdata != null) {
+            Material material = Material.getMaterial(materialdata);
+            if (material != null && material.isBlock()) {
+                display.data = material.getData();
+            }
```

Comment:
What's the point of having 3 different options that all accept a material? Do ItemStack, BlockData and MaterialData make a difference in the visual representation of the particle?

Reply:
- ItemStack only matches ITEM_CRACK particle (named "item" in vanilla).
- BlockData matches with BLOCK_CRACK (when a block is broken), BLOCK_DUST (When you run on a block) and FALLING_DUST (like falling sand particles below a flying sand block).
- MaterialData is for legacy usage, it's the magic id that modify the data. BlockData replaces MaterialData since 1.13 with a proper implementation, gread parameters (allowing us to use "stone_stairs[half=bottom,facing=north]" instead of "stone_stairs 7"). So, in 1.12 BLOCK_CRACK, BLOCK_DUST and FALLING_DUST are using MaterialData, but in 1.13 it's deprecated LEGACY_BLOCK_CRACK, LEGACY_BLOCK_DUST and LEGACY_FALLING_DUST.

In vanilla (or with Bukkit API), if one of this particles is used but no data is set, display fails. If one of this particles is used but a wrong data type is set, an error occurs.

I saw you support 1.12 (e.g. ISFLAT for DustOption) so I kept support for it even if users are in 1.13 or above. The best way is probaly to prevent users to set a data that doesn't matches the data type of the particle, but it needs to check the particle many times (on deserialization, when the particle changes, so accessors for particle field must be needed, and so on...).

My idea was to let users set a good type, knowing what he does, but if he uses a wrong type, it displays the particle without too many issues, with other parameters.

It's a kind of draft, it may be possible to set a Material in data field (no MaterialData, no BlockData, no ItemStack) and create the good type (MaterialData, BLockData or ItemStack) in spawns method depending on the particle, like for REDSTONE particle which uses a float[] instead of DustOption directly. But overload spawn method is possibly unsafe for performances, as it's the most frequently called method in XParticle.

Reply:
I see, that makes sense. Thanks for the explanation!